### DISCUSSION TITLE
Downgrade on upgrade

### DISCUSSION THREAD

**Discussion by hughsie** [fwupd/fwupd#6296] [Ideas]
We now have two situations where I think we might actually want to downgrade to a specific release when doing "fwupdmgr update":

* For some modems, replacing the SIM card means that we have to install a specific firmware for the correct carrier -- which might technically be a reinstall (same version) or even a downgrade -- the modem won't work until the new firmware is installed, which is why for modem firmware we already use `FU_DEVICE_INTERNAL_FLAG_SAVE_INTO_BACKUP_REMOTE` to save the .cab file locally as it's got all the carrier configs in. Fibocom has a userspace daemon and session process to do this right now, and it's _not awesome_ -- if we can do this in fwupd it would be much better.

* When devices have been flashed with the wrong firmware and we're *recovering* them to the correct stream/version we might want to install a firmware with the same version, but from a different GUID/requires stream.

I don't think we want to mix this into the Best Known Configuration stuff -- I think this would be per-device. I'm pondering a FUDevice internal flag called something like "`ENFORCE_SUPPORTED`" that would suggest a upgrade/downgrade/reinstall from `GetUpgrades()` to get the user back onto the latest LVFS-provided firmware. In event there was no available firmware we'd do nothing.

Insanity?

**Comment by superm1** [comment]
The problem with enforce supported is what happens when there really _IS_ a firmware upgrade available for the device?

* Can the device possibly know it's in a bad state and needs a reinstall/downgrade?  If so; I think it's better for it to "advertise" a lower version which forces the upgrade path.

**Reply by hughsie** [reply]
I guess an example might help -- The device I have here has GUID `USB\VID_413C&PID_81D7&REV_0318&CARRIER_Sprint` but installed with firmware `T77W968.F1.0.0.3.0.SB.005.016` designed for `USB\VID_413C&PID_81D7&REV_0318&CARRIER_SoftBank`. I want to install `T77W968.F1.0.0.4.6.SP.011` (from the same cab archive) when `CARRIER=Sprint` appears because I've inserted a new SIM card. Maybe this is a bad example as:

    rpmdev-vercmp "T77W968.F1.0.0.3.0.SB.005.016" "T77W968.F1.0.0.4.6.SP.011"
    T77W968.F1.0.0.3.0.SB.005.016 < T77W968.F1.0.0.4.6.SP.011

..but in the general case it might be a reinstall, or downgrade -- it's just the specific version string firmware the carrier has "blessed" to be in the cab archive.

I think what makes sense is for fwupd to say "hey this device with `USB\VID_413C&PID_81D7&REV_0318&CARRIER_Sprint` has a version that doesn't match with *anything* that exists in the metadata" and then upgrade would offer the device to (in this case update, but maybe downgrade) to the newest firmware that *does* have a match for `USB\VID_413C&PID_81D7&REV_0318&CARRIER_Sprint`.

**Reply by superm1** [reply]
This feels like a round screw in a pentagon hole.  It can kinda fit but not perfectly.

The moment that you're talking about _blessed versions_ that's totally BKC..

**Reply by hughsie** [reply]
Pentagon hole seems about right! I'm very confused why the card firmware and carrier config are linked together -- it seems totally crazy to me.

**Reply by superm1** [reply]
IMO if there needs to be a linkage between firmware version and carrier we either need a hack that resets the firmware version to 0 (so it forces the upgrade) or need to express it as a BKC.

The BKC expression doesn't really take this into account right now, but I would think that it could be extended so that you get certain devices that are measured into the BKC by GUID.  So when you swap the SIM card your GUID changes and so your BKC tag also changes.  Then knowing the updated BKC tag you can refresh all the firmware in the system.

I would think it's basically that you add a new device flag into any device that should be used to influence a BKC tag.

**Reply by hughsie** [reply]
My struggle with BKC tags are that they're system wide -- so if you had two modems, one on a SoftBank SIM and one on a Sprint SIM I'm not sure what we'd do when syncing the BKC.

**Reply by hughsie** [reply]
I think what we want is a `fwupd_device_set_bkc(FwupdDevice, const gchar *bkc)` that has a _per-device_ BKC of `Sprint` and then when we do the loop in `fu_util_get_release_with_tag()` [called from `fwupdmgr sync-bkc`] we compare the release against the per-system BKC **and also** the per-device BKC. I think that works, if we say the host-bkc has priority over the device one. That would mean adding `Sprint` as a tag to the firmware component too. Sane?

**Reply by superm1** [reply]
Yes much more sane.

**Reply by hughsie** [reply]
I do wonder at what point a per-device BKC is actually a branch? Couldn't we use `fwupd_device_set_branch(device, "Sprint")` and then make `sync-bkc` to the right thing with branches?

**Reply by superm1** [reply]
very good point!

**Reply by hughsie** [reply]
How about https://github.com/fwupd/fwupd/pull/6334 ?

**Reply by hughsie** [reply]
And then https://github.com/fwupd/fwupd/pull/6335 makes it useful.

**Reply by aleksander0m** [reply]
> I'm very confused why the card firmware and carrier config are linked together -- it seems totally crazy to me.

It is because carriers will certify with the FCC a very specific pair of firmware version and carrier config. Sometimes a given firmware version is expected in one single carrier, e.g. if they have changes only applicable to that single carrier.

**Reply by aleksander0m** [reply]
> I guess an example might help -- The device I have here has GUID USB\VID_413C&PID_81D7&REV_0318&CARRIER_Sprint but installed with firmware T77W968.F1.0.0.3.0.SB.005.016 designed for USB\VID_413C&PID_81D7&REV_0318&CARRIER_SoftBank. 

The CARRIER tag is (should be) loaded by ModemManager based on the currently selected carrier config. If the firmware version reports SB, I would have expected the CARRIER tag to be SoftBank instead. If this is not happening, it may be a bug in ModemManager. Newer Foxconn devices do that carrier config selection based on SIM card automatically, so it could also be that they are doing something wrong there, e.g. like not notifying that a carrier config switch actually happened. I believe there's an open issue where that exact same scenario is being discussed with Foxconn.

But I think I get the idea. You want the user to manually allow specifying which carrier they want to "follow" during a upgrade or downgrade operation, regardless of the current config they have selected, right?

**Reply by hughsie** [reply]
> You want the user to manually allow specifying which carrier they want to "follow" during a upgrade or downgrade operation

Kinda, the big picture is "user removes sim card, inserts new sim card" -> fwupd automatically installs the right firmware.

**Reply by aleksander0m** [reply]
In that case, after inserting the new SIM card, two things may happen:
 * if the carrier config for the new operator is already installed, MM autoselects it. The "upgrade path" of the device should now sync with the latest firmware version for that carrier.
 * if there is no installed carrier config for the new operator, MM should have switched automatically to the "GC" (generic) carrier config, and that would change the "upgrade path" of the device to "GC". All firmware versions shipped in LVFS should provide a GC variant (ideally). If there is a newer firmware update for GC, and it's installed by fwupd, after the installation MM may actually auto switch to the new operator carrier config automatically, if it was newly installed with the latest GC update.

all this is a bit tricky I guess. The key thing is that a cab in LVFS for a given firmware version should contain carrier configs for all operators using that specific version plus the GC one.

Automatic downgrade to a different version if e.g. the latest carrier config for a given operator is provided only on an older firmware version isn´t supported currently by the logic above. Your new approach may indeed help in that case, as the user could flag the device with a given carrier label, and request to sync to the latest best known version for that carrier explicitly.

What I don´t fuly understand is how the "use-branch" flag works, is it something that the user will manually set when they want to override the default behavior?

**Reply by hughsie** [reply]
> All firmware versions shipped in LVFS should provide a GC variant

I think that's where I'm stuck. Does the modem set the `mm_modem_get_carrier_configuration` in that scenario to `GC` or the actual carrier -- e.g. `VODAPHONE`? I suspect I'm using a modem without a GC config.

> What I don´t fuly understand is how the "use-branch" flag works

This would be set for "new" modems that are added to the modem-manager.quirk file without existing LVFS firmware -- I didn't want to break existing modems/firmware.

**Reply by aleksander0m** [reply]
> I think that's where I'm stuck. Does the modem set the mm_modem_get_carrier_configuration in that scenario to GC or the actual carrier -- e.g. VODAPHONE? I suspect I'm using a modem without a GC config.

Upon inserting the new SIM card, MM will try to find a matching carrier config for that SIM card. If there is no installed carrier config for the SIM right now, it should fallback to the generic one, see https://gitlab.freedesktop.org/mobile-broadband/ModemManager/-/blob/main/src/mm-shared-qmi.c#L2632.

When MM triggers the carrier config change, the whole modem object is reprobed, so the old modem object with the old carrier info goes away from DBus, and a new modem object is created with the new carrier info.

> This would be set for "new" modems that are added to the modem-manager.quirk file without existing LVFS firmware -- I didn't want to break existing modems/firmware.

I see. I think this makes sense. I assume a single .cab file in the LVFS can be tagged with multiple branch values, right?

**Reply by hughsie** [reply]
> I assume a single .cab file in the LVFS can be tagged with multiple branch values, right?

The cab archive can contain multiple firmwares, but each firmware can only have one branch right now. Are you thinking that some firmwares might need to have multiple branches? e.g. "VODAPHONE" **or** "SPRINT"?

**Reply by aleksander0m** [reply]
More or less, that is what we do with the DW5821e.  The same CAB file has one single system firmware image plus multiple carrier config images (when the CAB is installed, everything gets installed). This is like having multiple firmware versions provided by the same CAB file, because the firmware version gets composed with the version of the system firmware plus the carrier id plus the version of the carrier firmware, something along those lines.

**Comment by superm1** [comment]
This BKC usage does bring up another thought though - BKCs so far are mostly used for managed corporate use, right?  This is exposing BKCs into general purpose consumer stuff too.  We might need to think about the implications for GUI clients when it comes to BKCs like this.

**Reply by hughsie** [reply]
> BKCs so far are mostly used for managed corporate use, right

Right, and there's a plan to use them in ChromeOS too for managed deployments -- e.g. schools.
